### PR TITLE
[WiP] Rail gun disguised as inducer

### PR DESCRIPTION
**Pull request by eeSPee**
## About The Pull Request

Traitor item; rail gun induced as an inducer. Can be loaded and launches iron rods at people. Deals damage and can embed.

## Why It's Good For The Game

Rail guns are cool.

### WIP
* [x]  Gun and Ammo
* [x]  Test functionaliy
* [ ]  Uplink item

## Changelog
:cl:
add: rail gun disguised as inducer
/:cl:

### CODE REVIEW

**Comment by zeskorion** [comment]
You know,this concept would work a lot easier as a pneumatic cannon subtype

**Comment by eeSPee** [comment]
> You know,this concept would work a lot easier as a pneumatic cannon subtype

Yes, but that would need a gas tank to launch out items (which is not at all how rail guns work) and you would need to make rods actually embed when thrown, making them quite annoying.

**Comment by Isy232** [comment]
My thoughts:

1: I would think it should do a lot of damage per shot, since its like a decently long rod and rebar tends to be about half an inch thick, equal to .50 cal but really long. Also have it slowly draw power out of the cell in it into the "supercapacitor" on the gun or something. Just make it do large amounts of damage with the downside of a fairly long cooldown between shots.

2: Make it be loaded with standard power cells, if it is not already. Maybe have it consume 5 or 10 MJ per shot. You can expect sci mcguy or engi man to carry a lot of cells on them.

3: Possible engi/sci or just engi traitor role restriction.

4: If not already, massive armour penetration.

5: Maybe make it shoot sparks out when firing, I'd imagine it would.

_____

One thought would be that it could consume 10 MJ per shot, do 40 damage, 100% embed chance, large armour penetration, but with a 5-7 second cooldown between shots.

Or maybe 2.5 MJ per shot, 20 damage, 100% embed and large armour penetration still but only a 2 second cooldown between shots.

Maybe if you're insane have it consume 20 MJ, do 60 damage, still 100% embed and also throw people back, impaling them to a wall, or even another person if one if behind them. I'd imagine that would be a pain in the ass to code though. I'd think 10-12 second cooldown for this.

I dunno, the first sounds like a nieche weapon that would have its uses in prolonged fights where energy guns have run out of ammo but would be beaten if sec mcsecson was a good shot with a disabler and you lacked stims. It could also be a great hit and run weapon if you had a mobility advantage, such as a jetpack in space or meth in the halls.

The second sounds like a shittier revolver thats good against people with lots of armour and a lot more reusable. Good for a budget weapon if you don't want the steckin and have access to a lot of power cells.

The third sounds cool but maybe too OP.

Overall I think my first balance idea is best, it would be interesting for engi man to have a good role restricted weapon to make up for their general lack of any good standardish weapons from their base role.

___

Thats just my suggestions, you're coding it, take it wherever you want.

**Comment by eeSPee** [comment]
> 1: I would think it should do a lot of damage per shot, since its like a decently long rod and rebar tends to be about half an inch thick, equal to .50 cal but really long. Also have it slowly draw power out of the cell in it into the "supercapacitor" on the gun or something. Just make it do large amounts of damage with the downside of a fairly long cooldown between shots.

7 second might work. It might compensate for the easily acquired ammo (iron rods are easy to make and acquire anywhere on the staiton).

I initially thought of it as a ~3 second fire interval gun, because it can eat through ammo pretty fast, and it's not hard to avoid (compared to a spray and pray fully automatic) or you get penalized too much when you miss a shot.

> 2: Make it be loaded with standard power cells, if it is not already. Maybe have it consume 5 or 10 MJ per shot. You can expect sci mcguy or engi man to carry a lot of cells on them.

I liked the limitations of both being a energy gun (ie having to recharge at certain places, cell cannot be ejected) and the need for physical ammo as both drawbacks to it being a fairly decent gun available to traitors for TC, that can be recharged with station materials.

I wouldn't like a constantly draining power cell on a gun. It can compensate for the replaceable power cell on an energy gun, but it is just gonna be a hassle to deal with.

> 3: Possible engi/sci or just engi traitor role restriction.

No way around it.

>Maybe if you're insane have it consume 20 MJ, do 60 damage, still 100% embed and also throw people back, impaling them to a wall, or even another person if one if behind them. I'd imagine that would be a pain in the ass to code though. I'd think 10-12 second cooldown for this.

Zesko would never allow that :P

But that would be a cool idea for an advanced rail gun available for nukies.

File: code/modules/projectiles/guns/energy/special.dm [hunk 1]
```diff
@@ -364,3 +364,128 @@
...
+				embed_chance = 0,
+                embedded_fall_chance = 20,
+                embedded_ignore_throwspeed_threshold = TRUE
+			)
+			human_target.embed_object(embed_rod)
+			if (dropped && rod_drop)
+				qdel(rod_drop)
+
+/obj/item/projectile/bullet/reusable/railgun_rod/handle_drop()
+	if(dropped == FALSE)
```
Comment (utilbird) [inline code/modules/projectiles/guns/energy/special.dm]:
```suggestion
	if(!dropped)
```

File: code/modules/projectiles/guns/energy/special.dm [hunk 2]
```diff
@@ -364,3 +364,128 @@
...
+		var/mob/living/ltarget = target
+		if (istype(ltarget))
+			var/atom/throw_target = get_edge_target_turf(ltarget, angle2dir(Angle))
+			ltarget.throw_at(throw_target, 4, 2)
+		
+// --- THE GUN ---
+
+/obj/item/gun/energy/railgun
+	icon_state = "export_cannon"
+	name = "rail gun"
```
Comment (utilbird) [inline code/modules/projectiles/guns/energy/special.dm:423]: Name is usually above everything else

File: code/modules/projectiles/guns/energy/special.dm [hunk 3]
```diff
@@ -364,3 +364,128 @@
...
+	var/rod_loaded = TRUE
+
+/obj/item/gun/energy/railgun/update_icon(force_update)
+	if (!cell && cell.charge<1000)
+		icon_state = "export_cannon_empty"
+	else 
+		icon_state = "export_cannon"
+
+/obj/item/gun/energy/railgun/attack_self(mob/living/user as mob)
+	if(rod_loaded > 0)
```
Comment (utilbird) [inline code/modules/projectiles/guns/energy/special.dm]:
```suggestion
	if(rod_loaded)
```

File: code/modules/mob/living/carbon/human/human_defense.dm [hunk 4]
```diff
@@ -149,22 +149,24 @@
...
 				if(prob(I.embedding.embed_chance) && !HAS_TRAIT(src, TRAIT_PIERCEIMMUNE))
-					throw_alert("embeddedobject", /obj/screen/alert/embeddedobject)
-					var/obj/item/bodypart/L = pick(bodyparts)
-					L.embedded_objects |= I
-					I.add_mob_blood(src)//it embedded itself in you, of course it's bloody!
-					I.forceMove(src)
-					L.receive_damage(I.w_class*I.embedding.embedded_impact_pain_multiplier)
-					visible_message("<span class='danger'>[I] embeds itself in [src]'s [L.name]!</span>","<span class='userdanger'>[I] embeds itself in your [L.name]!</span>")
-					SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "embedded", /datum/mood_event/embedded)
+					src.embed_object(I)
```
Comment (utilbird) [inline code/modules/mob/living/carbon/human/human_defense.dm:155]: Doesn't this do the same without `src.` 🤔

**Review by utilbird — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/energy/special.dm [hunk 5]
```diff
@@ -364,3 +364,128 @@
...
+	
+/obj/item/gun/energy/railgun/recharge_newshot(no_cyborg_drain)
+	if (!rod_loaded)
+		return FALSE
+	. = ..()
+
+/obj/item/gun/energy/railgun/process_fire(atom/target, mob/living/user, message = TRUE, params = null, zone_override = "", bonus_spread = 0)
+	. = ..()
+	if (.)
+		rod_loaded=false
```
Comment (utilbird) [inline code/modules/projectiles/guns/energy/special.dm:464]: bools have to be fully capitalized, and it's good practice to put a space between the `=`

**Pull request closed without merging**